app/database2.js

```diff
@@ -0,0 +1,306 @@
+		getRecentDataset : "SELECT * FROM tbldatasets WHERE dataset_id=? AND ? > ?",
+    getDatasetByName : "SELECT * FROM tbldatasets WHERE dataset_name=?",
+		getDatasetItems : "SELECT dataset_items FROM tbldatasets WHERE dataset_id=?" ,
+		getGUILanguages : "SELECT * FROM tbllanguages WHERE language_gui = 1",
+		getUserSubjects : "SELECT * FROM tblsubjects",
+    getUser : "SELECT * FROM tblusers where user_id= ? ",
+    getUserbyEmail : "SELECT * FROM tblusers where user_email= ?",
+    getUserbyUsername : "SELECT * FROM tblusers where user_name= ?",
+		getUserIdbyUsername : "SELECT user_id, user_password FROM tblusers WHERE user_name=?",
+		getUserIdbyEmail : "SELECT user_id FROM tblusers WHERE user_email=?",
+    getLanguages: "SELECT * FROM tbllanguages",
+		getModules: "SELECT language_id, language_name, subject_id, subject_name FROM tbldatasets,tbllanguages,tblsubjects WHERE dataset_language=language_id AND dataset_subject=subject_id",
+		replaceUser: "REPLACE INTO tblusers VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
+		deleteDatasetbyId: "DELETE FROM tbldatasets WHERE dataset_id= ?",
+		deleteDatasetItemsbyId: "DELETE FROM tblitems WHERE item_dataset_id=?",
+		lastInsertedId: "SELECT ? FROM ? ORDER BY ? DESC LIMIT 1",
+		deleteDatasets: "DELETE FROM tbldatasets WHERE 1",
+		getSubjectByName : "SELECT * FROM tblsubjects WHERE subject_name=?"
+	};
+
+	var lastId = 0;
+
+	// Check if SQL.js has been loaded through AMD
+	var sql,db,db_online;
+	if (typeof sqlite !== 'object') {
+		document.body.style.backgroundColor = 'red';
+		alert("Failed to require sqlite through AMD.");
+	} else {
+		sql = sqlite;
+	}
+
+	// Initiate DB and check if there is an existing user DB
+	var read_database;
+	if (database_exists(config.constant("DATABASE_USER"))) {
+		read_database = fs.readFileSync(config.constant("DATABASE_USER"));
+	} else {
+		read_database = fs.readFileSync(config.constant("DATABASE_SLIMSTAMPEN"));
+	}
+	var db = new sql.Database(read_database);
+
+	function database_exists(path) {
+		try {
+			fs.accessSync(path, fs.F_OK);
+			return true;
+		} catch (e) {
+			return false;
+		}
+	}
+
+	function onError(error) {
+		if (error) {
+			if (error.message !== undefined) {
+				console.log(error.message);
+			} else {
+				console.log(error);
+			}
+		}
+	}
+	function isUnique(unique_name1, unique_name2, queryResult, row) {
+	    for (i = 0; i<queryResult.length;i++) {
+	      if (queryResult[i][unique_name1]==row[unique_name1] && queryResult[i][unique_name2]==row[unique_name2]) {
+					return false;
+	      }
+	    }
+	    return true;
+	}
+  function isUniqueOld(unique_name, queryResult, row) {
+    for (i = 0; i<queryResult.length;i++) {
+      if (queryResult[i][unique_name]==row[unique_name]) {
+        return false;
+      }
+    }
+    return true;
+  }
+	function synchronizeUser(userId, callback) {
+		var local_user = database.getQuery('getUser',[userId]);
+		database.getOnlineQuery("getUser", [userId], function(online_user) {
```

Comment:
This function is way to big. There is too much on too many different abstractions levels. And the conditions you use, e.g. `key=="user_createdate" || key=="user_lastedited" || key=="user_bday"`aren't very clear.